src/protocols/secure_channel/CASESession.cpp

```diff
@@ -150,18 +150,19 @@ class CASESession::WorkHelper
     // The `status` value is the result of the work callback (called beforehand).
     typedef CHIP_ERROR (CASESession::*AfterWorkCallback)(DATA & data, CHIP_ERROR status);
 
-    // Create a work helper using the specified session, work callback, after work callback, and data (template arg).
-    // Lifetime is not managed, see `Create` for that option.
-    WorkHelper(CASESession & session, WorkCallback workCallback, AfterWorkCallback afterWorkCallback) :
-        mSession(&session), mWorkCallback(workCallback), mAfterWorkCallback(afterWorkCallback)
-    {}
-
+public:
     // Create a work helper using the specified session, work callback, after work callback, and data (template arg).
     // Lifetime is managed by sharing between the caller (typically the session) and the helper itself (while work is scheduled).
     static Platform::SharedPtr<WorkHelper> Create(CASESession & session, WorkCallback workCallback,
                                                   AfterWorkCallback afterWorkCallback)
     {
-        auto ptr = Platform::MakeShared<WorkHelper>(session, workCallback, afterWorkCallback);
+        struct EnableShared : public WorkHelper
```

Comment:
Oh, this is cute.  I should remember this trick, with the completely hidden struct definition and all.

Reply:
Yeah had to look this one up. It's easier if it's only default constructors. And there's also a way to use a template to generate any constructor (with perfect forwarded args) but not if it's an inner class, it would have to be more standalone.

Reply:
Note that the deleter gets the correct destructor even if it's not virtual. (I double checked.)